src/codeeditor/editor.ts

```diff
@@ -0,0 +1,637 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  find
+} from 'phosphor/lib/algorithm/searching';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  ISignal, clearSignalData, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  IChangedArgs
+} from '../common/interfaces';
+
+import {
+  ObservableVector
+} from '../common/observablevector';
+
+import {
+  JSONObject
+} from 'phosphor/lib/algorithm/json';
+
+
+/**
+ * A namespace for code editors.
+ */
+export
+namespace CodeEditor {
+  /**
+   * A zero-based position in the editor.
+   */
+  export
+  interface IPosition {
+    /**
+     * The cursor line number.
+     */
+    line: number;
+
+    /**
+     * The cursor column number.
+     */
+    column: number;
+  }
+
+  /**
+   * An interface describing editor state coordinates.
+   */
+  export
+  interface ICoords {
+    /**
+     * The left coordinate value.
+     */
+    readonly left: number;
+
+    /**
+     * The right coordinate value.
+     */
+    readonly right: number;
+
+    /**
+     * The top coordinate value.
+     */
+    readonly top: number;
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * A text selection.
+   */
+  export
+  interface ITextSelection {
+    /**
+     * The index to the first character in the current selection.
+     *
+     * #### Notes
+     * If this position is greater than [end] then the selection is considered
+     * to be backward.
+     */
+    start: number;
+
+    /**
+     * The index to the last character in the current selection.
+     *
+     * #### Notes
+     * If this position is less than [start] then the selection is considered
+     * to be backward.
+     */
+    end: number;
+
+    /**
+     * The uuid of the text selection owner.
+     */
+    uuid: string;
+
+    /**
+     * A class name added to the text selection.
+     */
+    className?: string;
+  }
+
+  /**
+   * An editor model.
+   */
+  export
+  interface IModel extends IDisposable {
+    /**
+     * A signal emitted when the value changes.
+     */
+    valueChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * A signal emitted when a property changes.
+     */
+    mimeTypeChanged: ISignal<IModel, IChangedArgs<string>>;
+
+    /**
+     * The text stored in the model.
+     */
+    value: string;  // TODO: this should be an iobservablestring.
+
+    /**
+     * A mime type of the model.
+     */
+    mimeType: string;
+
+    /**
+     * Get the number of lines in the model.
+     */
+    readonly lineCount: number;
+
+    /**
+     * The currently selected code.
+     *
+     * @returns A read-only copy of the text selections.
+     */
+    readonly selections: ObservableVector<ITextSelection>;
```

Comment:
@blink1073 I've prototyped a bit with selections and have the following suggestions:
- remove `uuid` from `ITextSelection`
- introduce `IModelOwner`:
```ts 
  /**
   * An editor model owner.
   */
  export
  interface IModelOwner {
    /**
     * The uuid of the owner.
     */
    readonly uuid: string;
    /**
     * The selections for all the cursors.
     * Cursors will be removed or added, as necessary.
     */
    readonly selections: ObservableVector<ITextSelection>;
    /**
     * Set the primary position of the cursor. This will remove any secondary cursors.
     */
    setPosition(position: IPosition): void;
    /**
     * Returns the primary position of the cursor.
     */
    getPosition(): IPosition;
    /**
     * Set the primary selection. This will remove any secondary cursors.
     */
    setSelection(selection: ITextSelection): void;
    /**
     * Returns the primary selection.
     */
    getSelection(): ITextSelection;
  }
```
- replace `IModel.selections` property with `IModel.owners`:
```ts
/**
 * The owners of the current model.
 */
readonly owners: ObservableVector<IModelOwner>;
```
- Provide a default impl `IModelOwner` which just a wrapper around `selections` - `ModelOwner`
- Make `IEditor` extends `IModelOwner`
- Implementations of IEditor should:
  - create an instance of `ModelOwner` and add as an owner to `IModel.owners`
  - implement `IModelOwner` by delegating to an instance of `ModelOwner`
  - subscribe to changes of owners and changes of their selections and manage cursors based on it
- to add foreign cursors or selections one can create an instance of `ModelOwner` and add it to `IModel.owners`

What do you think?